site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/repository/HapiFhirRepository.java

```diff
@@ -37,92 +37,138 @@
+     *
+     * @param resource the resource to create
+     * @return id of the created resource
      */
-    public String saveOrganization(Organization organization) {
-        logger.info(FhirRepo.SAVE_REQUEST.message(),
-                fhirContextWrapper.prettyPrint(organization));
-
+    private String saveResource(final Resource resource) {
+        logger.info(Repositorys.REQUEST_PAYLOAD.message(), fhirContextWrapper.prettyPrint(resource));
         Bundle responseBundle;
         try {
             responseBundle = fhirContextWrapper.executeTransaction(fhirUri,
-                bundle(organization, ORGANIZATION_ENTITY_NAME));
-        } catch (BaseServerResponseException e) {
-            logger.warn(FhirRepo.UPDATE_ERROR.message(), e);
-            throw new RestException(e.getMessage(), e);
+                    bundle(resource, resource.getResourceType().name()));
+        } catch (FhirServerResponseException e) {
+            throw new RestException(Repositorys.UPDATE_ERROR.message(), e);
         }
         IBaseResource responseElement = extractResponseResource(responseBundle);
-
-        // Log the response
-        logger.info(FhirRepo.SAVE_RESPONSE.message(),
-                    fhirContextWrapper.prettyPrint(responseElement));
-
+        logger.info(Repositorys.RESPONSE_PAYLOAD.message(), fhirContextWrapper.prettyPrint(responseElement));
         return responseElement.getIdElement().getIdPart();
     }
 
     /**
-     * @param researchStudy the researchStudy to save.
-     * @return ResearchStudy
+     * @param organization the organization to save.
+     * @return id of the saved organization
      */
-    public String saveResearchStudy(ResearchStudy researchStudy) {
-        // Log the request
-
-        logger.info(FhirRepo.SAVE_REQUEST.message(),
-                    fhirContextWrapper.prettyPrint(researchStudy));
+    public String saveOrganization(final Organization organization) {
+        return saveResource(organization);
+    }
 
-        Bundle responseBundle;
+    /**
+     * Exact search on name
+     *
+     * @param name of the organization to search.
+     * @return org with that name or none() if not found
+     */
+    @Override
+    public Optional<Organization> findOrganizationByName(final String name) {
         try {
-            responseBundle = fhirContextWrapper.executeTransaction(fhirUri,
-                    bundle(researchStudy, RESEARCHSTUDY_ENTITY_NAME));
+            return fhirContextWrapper
+                    .search(fhirUri, Organization.class)
+                    .where(Organization.NAME.matchesExactly().value(name))
+                    .returnBundle(Bundle.class)
+                    .execute()
+                    .getEntry()
+                    .stream()
+                    .findFirst()
+                    .map(Bundle.BundleEntryComponent::getResource)
+                    .filter(r -> r.getResourceType().equals(ResourceType.Organization))
```

Comment:
are 
 .filter(r -> r.getResourceType().equals(ResourceType.Organization))
 .map(Organization.class::cast);
necessary? 
In.
.getEntry()
                    .stream()
                    .findFirst()
                    .map(Bundle.BundleEntryComponent::getResource)
                    .filter(r -> r.getResourceType().equals(ResourceType.Organization))
                    .map(Organization.class::cast);

wouldnt 
.getEntry()
                    .stream()
                    .findFirst()
                    .map(Organization.class::cast);
be ok?

Reply:
That would work if the response is guaranteed to contain only Organization, if not it would get the first resource and try to cast it to Organization, which might fail. I believe Tamir's new fhir-client library has a typesafe search method, so I would rather update when that is ready.

Reply:
I thought ..
.search(fhirUri, Organization.class)
 means we only get Organization?